README.md
Symfony JavaScript Router
=========================

This is a modern implementation of the JS router for generation Symfony routes.
Best to be used with the [JavaScript routing bundle](https://github.com/Becklyn/javascript-routing-bundle).


Initialization
--------------

For example like this (using `mojave`):

```js
import {initFromGlobalData} from "mojave/init"; 
import {Router} from "@becklyn/router";

let router = initFromGlobalData("RouterInit", new Router());
```


A manual way to wire it up to global data is like this: 

```js
function loadGlobalData (key, handler)
{
    if (undefined === window[key])
    {
        return;
    }

    // replace global callback
    window[key].init = (data) => handler.init(data);
    // handle current data
    handler.init(window[key].data);
}

let router = new Router();
loadGlobalData("RouterInit", router);
```


CHANGELOG.md
1.0.3
=====

*   (bug) Ensure that passed parameters are never modified.


1.0.2
=====

*   Add `module` key in `package.json`.
*   Remove dependency on `mojave`.


1.0.1
=====

*   Fixed internal types in the `Router`.


1.0.0
=====

Initial release `\o/`


tests/variations.js
import {
    runDifferentContextRouteTests,
    runDifferentParametersTests
} from "./helpers/test-runner";

// test simple cases

runDifferentParametersTests(
    "Simple tests",
    {
        baseUrl: "",
        host: "example.org",
        ports: {
            http: 80,
            https: 443,
        },
        scheme: "https",
    },
    {
        host: [],
        path: [["text", "/test"]],
        schemes: [],
        variables: [],
    },
    [
        {
            params: [],
            expected: "/test",
        },
        // Reference type => ABSOLUTE_URL
        {
            params: [{}, 0],
            expected: "https://example.org/test",
        },
        // Reference type => ABSOLUTE_PATH
        {
            params: [{}, 1],
            expected: "/test",
        },
        // Reference type => NETWORK_PATH
        {
            params: [{}, 3],
            expected: "//example.org/test",
        },
        // Fragment variations
        {
            params: [{_fragment: "abc"}],
            expected: "/test#abc",
        },
        {
            params: [{_fragment: 123}],
            expected: "/test#123",
        },
        {
            params: [{_fragment: true}],
            expected: "/test#1",
        },
        {
            params: [{_fragment: false}],
            expected: "/test",
        },
        {
            params: [{_fragment: null}],
            expected: "/test",
        },
        {
            params: [{_fragment: undefined}],
            expected: "/test",
        },
    ]
);

// test fully featured
runDifferentParametersTests(
    "Fully featured test",
    {
        baseUrl: "",
        host: "example.org",
        ports: {
            http: 80,
            https: 443,
        },
        scheme: "https",
    },
    {
        host: [],
        path: [
            ["text", "/suffix"],
            ["variable", "/", ".*", "test"],
            ["text", "/prefix"],
        ],
        schemes: [],
        variables: ["test"],
    },
    [
        {
            params: [{test: 123}],
            expected: "/prefix/123/suffix",
        },
        {
            params: [{test: 123, q1: "abc", q2: 123, q3: true, q4: false, q5: null, q6: undefined, _fragment: "abc"}],
            expected: "/prefix/123/suffix?q1=abc&q2=123&q3=1&q4=0#abc",
        },
    ]
);


runDifferentContextRouteTests(
    "Differences between definition and context",
    [
        // #1 base example
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 443,
                },
                scheme: "http",
            },
            route: {
                host: [],
                path: [
                    ["text", "/test"],
                ],
                schemes: [],
                variables: [],
            },
            expected: "/test",
        },
        // #2 different host
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 443,
                },
                scheme: "http",
            },
            route: {
                host: [
                    ["text", "other.example.org"],
                ],
                path: [
                    ["text", "/test"],
                ],
                schemes: [],
                variables: [],
            },
            expected: "//other.example.org/test",
        },
        // #3 different scheme (use first in list 1/2)
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 443,
                },
                scheme: "http",
            },
            route: {
                host: [],
                path: [
                    ["text", "/test"],
                ],
                schemes: ["https", "ftp"],
                variables: [],
            },
            expected: "https://example.org/test",
        },
        // #4 different scheme (use first in list 2/2)
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 443,
                },
                scheme: "http",
            },
            route: {
                host: [],
                path: [
                    ["text", "/test"],
                ],
                schemes: ["ftp", "https"],
                variables: [],
            },
            expected: "ftp://example.org/test",
        },
        // #5 different scheme + port
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 1443,
                },
                scheme: "http",
            },
            route: {
                host: [],
                path: [
                    ["text", "/test"],
                ],
                schemes: ["https"],
                variables: [],
            },
            expected: "https://example.org:1443/test",
        },
        // #5 different port
        {
            context: {
                baseUrl: "",
                host: "example.org",
                ports: {
                    http: 80,
                    https: 1443,
                },
                scheme: "https",
            },
            route: {
                host: [],
                path: [
                    ["text", "/test"],
                ],
                schemes: ["https"],
                variables: [],
            },
            expected: "//example.org:1443/test",
        },
    ],
    [],
);


tests/helpers/test-runner.js
import test from "ava";
import {Router} from "../../src/index";


/**
 * Runs a suite of tests, with different params / reference type combinations.
 *
 * Will create a route named `test` with the given route config and
 * call the router with the params in the tests and match it against
 * the expected value in the params
 *
 * @param {string} description
 * @param {BecklynJavaScriptRouter.Context} context
 * @param {BecklynJavaScriptRouter.Route} routeConfig
 * @param {Array<{params: Array<*>, expected: string}>} tests
 */
export function runDifferentParametersTests (description, context, routeConfig, tests)
{
    tests.forEach((testData, index) => {
        let router = new Router();
        router.init({
            context: context,
            routes: {
                test: routeConfig
            },
        });

        test(`test '${description}', #${index + 1}`, t => {
            let params = testData.params;
            t.is(router.generate("test", ...params), testData.expected);
        });
    });
}

/**
 * Runs a suite of tests, with different context / route combinations.
 *
 * @param {string} description
 * @param {Array<{context: BecklynJavaScriptRouter.Context, route: BecklynJavaScriptRouter.Route, expected: string}>} tests
 * @param {Array<*>} params
 */
export function runDifferentContextRouteTests (description, tests, params)
{
    tests.forEach((testData, index) => {
        let router = new Router();
        router.init({
            context: testData.context,
            routes: {
                test: testData.route
            },
        });

        test(`test '${description}', #${index + 1}`, t => {
            t.is(router.generate("test", ...params), testData.expected);
        });
    });
}


tests/immutable-parameters.js
import test from "ava";
import {Router} from "../src/index";


test("Test that parameters are not modified", t => {
    let params = {
        a: 1,
        b: 2,
        test: 123,
    };

    let router = new Router();
    router.init({
        context: {
            baseUrl: "",
            host: "example.org",
            ports: {
                http: 80,
                https: 443,
            },
            scheme: "https",
        },
        routes: {
            test: {
                host: [],
                path: [
                    ["variable", "/", ".*", "test"],
                    ["text", "/suffix"],
                ],
                schemes: [],
                variables: ["test"],
            }
        }
    });

    let url = router.generate("test", params);
    t.is(url, "/suffix/123?a=1&b=2");
    t.is(params.a, 1);
    t.is(params.b, 2);
    t.is(params.test, 123);
});


src/@types/router.d.ts
export module BecklynJavaScriptRouter
{
    type TextToken = [
        "text", // type
        string, // text content
    ];

    type VariableToken = [
        "variable", // type
        string,     // preceding text
        string,     // regex to match the variable
        string,     // variable name
    ];
    type Token = TextToken|VariableToken;

    interface PortMapping
    {
        [scheme: string]: number;
    }

    /**
     * The data of a single route
     */
    export interface Route
    {
        host: Token[];
        path: Token[];
        schemes: string[];
        variables: string[];
    }

    /**
     * The router context
     */
    export interface Context
    {
        baseUrl: string;
        host: string;
        ports: PortMapping;
        scheme: string;
    }

    /**
     * All the routes in the route collection
     */
    export interface RouteCollection
    {
        [name: string]: Route;
    }

    /**
     * The complete init data
     */
    export interface InitData
    {
        context: Context;
        routes: RouteCollection;
    }
}


src/index.ts
import {encode} from "qss";
import {BecklynJavaScriptRouter} from "./@types/router";

const DEFAULT_PORTS: BecklynJavaScriptRouter.PortMapping = {
    http: 80,
    https: 443,
};

/**
 *
 */
type ParameterType = string | number | boolean | null | undefined
export interface RouteParameters
{
    [parameter: string]: ParameterType;
}

export enum ReferenceType
{
    ABSOLUTE_URL = 0,
    ABSOLUTE_PATH = 1,
    // RELATIVE_PATH is not supported, as it is pretty niche and would add quite a lot of bytes to the build
    NETWORK_PATH = 3,
}

/**
 * Sanitizes the pattern so that it works in JS.
 *
 * Sanitized cases:
 *      * A double + (`++`, possessive +) isn't supported in JS, so we just use a single `+` instead.
 */
function sanitizeRegex (pattern: string) : string
{
    return pattern.replace(/\++/, "+");
}


/**
 * Compiles the tokens to a string
 */
function compileTokens (tokens: BecklynJavaScriptRouter.Token[], parameters: Readonly<RouteParameters>) : string
{
    let compiled = tokens.map(token =>
    {
        if (token[0] === "text")
        {
            return token[1];
        }

        let regex = token[2];
        let name = token[3];
        let value = stringifyValue(parameters[name]);
        let matcher = new RegExp(`^${sanitizeRegex(regex)}$`);

        if (!matcher.test(value))
        {
            throw new Error(`Parameter '${name}' must match '${regex}' ('${value}' given).`)
        }

        return token[1] + value;
    });

    return compiled.reverse().join("");
}

/**
 * Stringifies a value for usage as path / query / host parameter or fragment
 */
function stringifyValue (value: string|number|boolean|null|undefined, explicitFalse: boolean = false) : string
{
    if (true === value)
    {
        return "1";
    }

    if (false === value)
    {
        return explicitFalse ? "0" : "";
    }

    return null != value
        ? value.toString()
        : "";
}

/**
 * Generates the query string
 */
function generateQuery (variables: string[], parameters: Readonly<RouteParameters>) : string
{
    let usedVariables: {[name: string]: boolean} = {};

    // Mark all used variables
    variables.forEach(name => usedVariables[name] = true);

    // filter all null / undefined parameters
    let filtered: RouteParameters = {};
    for (let name in parameters)
    {
        if (!usedVariables[name] && null != parameters[name] && "_fragment" !== name)
        {
            filtered[name] = stringifyValue(parameters[name], true);
        }
    }

    return Object.keys(filtered).length > 0
        ? encode(filtered, "?")
        : "";
}

/**
 * Generates the fragment part
 */
function generateFragment (parameters: Readonly<RouteParameters>) : string
{
    let fragment = parameters["_fragment"];

    return !fragment
        ? ""
        : "#" + stringifyValue(fragment);
}

/**
 * Router, that generates URLs
 */
export class Router
{
    private context: BecklynJavaScriptRouter.Context = {
        baseUrl: "",
        host: "",
        ports: DEFAULT_PORTS,
        scheme: "https",
    };
    private routes: BecklynJavaScriptRouter.RouteCollection = {};


    /**
     * Initializes the router
     */
    public init (data: BecklynJavaScriptRouter.InitData)
    {
        this.context = data.context;
        this.routes = data.routes;
    }


    /**
     * Generates a URL or path for a specific route based on the given parameters.
     */
    public generate (name: string, parameters: Readonly<RouteParameters> = {}, referenceType: ReferenceType = ReferenceType.ABSOLUTE_PATH) : string
    {
        let route = this.routes[name];

        //region Check for missing route
        if (undefined === route)
        {
            throw new Error(`Route not found: '${name}'`);
        }
        //endregion


        //region Check for missing parameters
        let missingParameters = route.variables.filter(
            variable => {
                return undefined === parameters[variable];
            }
        );

        if (missingParameters.length > 0)
        {
            let paramNames = Object.keys(parameters);
            throw new Error(`Failed to generate route '${name}': Some mandatory parameters are missing or undefined ("${missingParameters.join('", "')}"). Provided were ${paramNames.length ? paramNames.join(", ") : "no parameters"}`);
        }
        //endregion

        try
        {
            let host = compileTokens(route.host, parameters);
            let segments: string[] = [];
            let isMissingRequiredScheme = route.schemes.length > 0 && -1 === route.schemes.indexOf(this.context.scheme);
            let includeScheme = false;
            let scheme = isMissingRequiredScheme
                ? route.schemes[0]
                : this.context.scheme;
            let usedPort = this.context.ports[scheme];
            let needsExplicitPort = usedPort !== undefined && usedPort !== DEFAULT_PORTS[scheme];

            if (referenceType === ReferenceType.ABSOLUTE_URL || isMissingRequiredScheme)
            {
                segments.push(`${scheme}:`);
                includeScheme = true
            }

            if (
                (referenceType === ReferenceType.NETWORK_PATH || includeScheme)
                || ("" !== host && host !== this.context.host)
                || needsExplicitPort
            )
            {
                segments.push("//");
                segments.push(host || this.context.host);

                if (needsExplicitPort)
                {
                    segments.push(":" + usedPort);
                }
            }

            segments.push(this.context.baseUrl);
            segments.push(compileTokens(route.path, parameters));
            segments.push(generateQuery(route.variables, parameters));
            segments.push(generateFragment(parameters));

            return segments.join("");
        }
        catch (e)
        {
            throw new Error(`Failed to generate route '${name}': ${e.message}`);
        }
    }
}
